### PR TITLE
8423 Modular plugins

### PR DESCRIPTION
**Pull request by alexander-fedorenko**
## Description
This PR introduces support of module plugins.
See #8423 and https://github.com/geosolutions-it/MapStore2/wiki/%5BProposal%5D-MapStore-Modular-Plugins
for more details.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [x] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**
#8423

**What is the new behavior?**
MapStore supports module plugins. Plugin implementation is loaded dynamically on demand.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/containers/MapViewer.jsx
Comment (allyoucanmap) [inline web/client/containers/MapViewer.jsx:21]: We can remove compose now

File: web/client/containers/Page.jsx
Comment (allyoucanmap) [inline web/client/containers/Page.jsx:19]: We can remove compose now

File: web/client/utils/StateUtils.js
Comment (allyoucanmap) [inline web/client/utils/StateUtils.js:118]: Could we remove this including the relative doc?

File: web/client/utils/StateUtils.js
Comment (allyoucanmap) [inline web/client/utils/StateUtils.js:204]: Is this code still needed?

File: web/client/utils/StateUtils.js
Comment (allyoucanmap) [inline web/client/utils/StateUtils.js:294]: It is not clear how this `notMutable` key works to me.
Is it working because there is not a plugin called notMutable so the mute and unmute handler will not be triggered?

File: web/client/product/components/viewer/MapViewerCmp.jsx
Comment (allyoucanmap) [inline web/client/product/components/viewer/MapViewerCmp.jsx:51]: Based on the pvt discussion we should do the follow:
- revert this fix for the moment with all related code around
- plugin that has still issue because of this should be kept static for the moment
- once we have the list with all the problematic plugins we can evaluate how to proceed

File: web/client/product/plugins.js
Comment (allyoucanmap) [inline web/client/product/plugins.js:10]: Please replace require with import for static plugins

**Review by allyoucanmap — COMMENTED** [review]
The implementation seems good, most of the comments are related to naming, questions or typos.
Before to proceed with the issue related with specific plugin I would suggest to complete the support for extension, once we have also that support I'll do another round of test.

We should also verify the final bundles sizes with the webpack analyzer after this initial work

File: web/client/components/app/main.jsx
Comment (alexander-fedorenko) [reply web/client/components/app/main.jsx]: Removed

File: web/client/components/app/withExtensions.js
Comment (alexander-fedorenko) [reply web/client/components/app/withExtensions.js]: Reverted this change, thanks

File: web/client/components/plugins/PluginsContainer.jsx
Comment (alexander-fedorenko) [reply web/client/components/plugins/PluginsContainer.jsx]: Did you mean we should not add? I removed connect and moved it instead into HOC.
Export with HOC allows to simplify usage of resulting component on a list of pages. In a few places import + usage of connect is replaced now with only import; another places doing connect to a fewer list of state values. It also makes code a bit more error-proof.

File: web/client/containers/MapViewer.jsx
Comment (alexander-fedorenko) [reply web/client/containers/MapViewer.jsx]: Sure, removed

File: web/client/utils/StateUtils.js
Comment (alexander-fedorenko) [reply web/client/utils/StateUtils.js]: Removed

File: web/client/utils/StateUtils.js
Comment (alexander-fedorenko) [reply web/client/utils/StateUtils.js]: Removed commented line

File: web/client/components/plugins/PluginsContainer.jsx
Comment (allyoucanmap) [reply web/client/components/plugins/PluginsContainer.jsx]: Why do we need the connect also in the HOC?
We are already connecting everything in the page, both hook, HOC and component should not have connect

File: web/client/utils/StateUtils.js
Comment (alexander-fedorenko) [reply web/client/utils/StateUtils.js]: Exactly, epics are mutable by "notMutable" key, and this will never happen because there is no such plugin.
But augmentStore was used only by withExtensions HOC. With the recent commits this HOC was slightly updated and now it uses storeManager.addEpics & storeManager.addReducer methods to manager reducers and epics.

File: web/client/product/components/viewer/MapViewerCmp.jsx
Comment (alexander-fedorenko) [reply web/client/product/components/viewer/MapViewerCmp.jsx]: Rolled back, list of plugins is to be added in the comment down below in PR.

File: web/client/product/plugins.js
Comment (alexander-fedorenko) [reply web/client/product/plugins.js:10]: Replaced

**Comment by alexander-fedorenko** [comment]
Following plugins are kept as static right now (plugins name is followed by issue description):

-  ContextPlugin - no data is visible on context page. This is a flow issue, required action or actions are dispatched before plugin epics are registered and ready to listen for actions.
-  Dashboard - same as above.
-  DashboardsPlugin - same as above.
-  GeoStoryPlugin - same as above.
-  GeoStoriesPlugin - same as above.
-  MapsPlugin - same as above.
-  FeedbacMaskPlugin - rely on INIT_MAP action that is dispatched before pluing epics are registered.
-  PrintPlugin - probably a conflict with how plugin is defined. Need to be investigated.
-  UserSession - need investigation.

Following plugins are related to GeoFence integration and will require specific environment to test.
- RulesDataGridPlugin,
-  RulesEditorPlugin,
-  RulesManagerFooter,

This one needs specific layers available to test:
-  ThematicLayerPlugin,

Is this deprecated?
-  StyleEditorPlugin.

File: web/client/components/plugins/PluginsContainer.jsx
Comment (alexander-fedorenko) [reply web/client/components/plugins/PluginsContainer.jsx]: As agreed in a call, applied solution with having separate container "ModulePluginsContainer" specifically for MapStore pages. This container is wrapped by a "withModulePlugins" HOC to support processing of module plugins.

**Comment by tdipisa** [comment]
@allyoucanmap let's evaluate/check the effective benefits we have now in performances before moving further steps and update plugins in the list above. Let's proceed step by step. I would like to understand how/when we have already improved the loading time to better schedule organize other steps accordingly.

**Comment by allyoucanmap** [comment]
# Bundle Analyzer

processed with `webpack-bundle-analyzer`

## Parsed

mapstore2.js from 9.53 MB (bc98bd2737c35794b94fd7dc442502cb032c2018) to 6.75 MB ([alexander-fedorenko:523-modular-plugins](https://github.com/alexander-fedorenko/MapStore2/tree/523-modular-plugins))

image
image

## Gzipped

mapstore2.js from 2.46 MB (bc98bd2737c35794b94fd7dc442502cb032c2018) to 1.76 MB ([alexander-fedorenko:523-modular-plugins](https://github.com/alexander-fedorenko/MapStore2/tree/523-modular-plugins))

image
image


# Excluded static plugins test

This tests removes the static plugins to verify what should be the main bundle size if all the plugins are imported as module

image

## Parsed

mapstore2.js from 9.53 MB (bc98bd2737c35794b94fd7dc442502cb032c2018) to 2.36 MB

image

## Gzipped

mapstore2.js from 2.46 MB (bc98bd2737c35794b94fd7dc442502cb032c2018) to 639.36 KB

image

The interesting part of this test is the content of mapstore2.js without static plugins included because it highlight library that should be core. All seems fine apart from `css-tree` that is related to annotation/marker glyphs and we could probably find a way to load this library only when it's really needed (currently we load everything at the start)

image

**Comment by allyoucanmap** [comment]
> Following plugins are kept as static right now (plugins name is followed by issue description):
> 
> -  ContextPlugin - no data is visible on context page. This is a flow issue, required action or actions are dispatched before plugin epics are registered and ready to listen for actions.
> -  Dashboard - same as above.
> -  DashboardsPlugin - same as above.
> -  GeoStoryPlugin - same as above.
> -  GeoStoriesPlugin - same as above.
> -  MapsPlugin - same as above.
> -  FeedbacMaskPlugin - rely on INIT_MAP action that is dispatched before plugin epics are registered.
> -  PrintPlugin - probably a conflict with how plugin is defined. Need to be investigated.
> -  UserSession - need investigation.

let's wait for this at the moment

> Following plugins are related to GeoFence integration and will require specific environment to test.
> - RulesDataGridPlugin,
> -  RulesEditorPlugin,
> -  RulesManagerFooter,

> This one needs specific layers available to test:
> -  ThematicLayerPlugin,
> 

@tdipisa do you know a project that is using these plugins?

> Is this deprecated?
> -  StyleEditorPlugin.

It's not deprecated it should work if you have admin user and a geoserver configured to edit styles, it should work also as module plugin (but I did not tested)

image

Note:

We are currently using the [initial state configuration](https://mapstore.readthedocs.io/en/latest/developer-guide/local-config/#initialstate-configuration) to apply some configuration at start up for following plugins.
We need to verify if this configuration is still valid if the plugins are loaded after the initialization of standard app

- Catalog 
- Measure
- Annotations

```js
{
    "initialState": {
        "defaultState": {
            "annotations": {
                "config": {
                    "multiGeometry": true,
                    "validationErrors": {}
                },
                "defaultTextAnnotation": "New"
            },
            "maptype": {
                "mapType": "{context.mode === 'desktop' ? 'openlayers' : 'leaflet'}"
            },
            "catalog": {
                "default": {
                    "newService": {
                        "url": "",
                        "type": "wms",
                        "title": "",
                        "isNew": true,
                        "autoload": false
                    },
                    "selectedService": "gs_stable_csw",
                    "services": {
                        "gs_stable_csw": {
                            "url": "https://gs-stable.geo-solutions.it/geoserver/csw",
                            "type": "csw",
                            "title": "GeoSolutions GeoServer CSW",
                            "autoload": true
                        },
                        "gs_stable_wms": {
                            "url": "https://gs-stable.geo-solutions.it/geoserver/wms",
                            "type": "wms",
                            "title": "GeoSolutions GeoServer WMS",
                            "autoload": false
                        },
                        "gs_stable_wmts": {
                            "url": "https://gs-stable.geo-solutions.it/geoserver/gwc/service/wmts",
                            "type": "wmts",
                            "title": "GeoSolutions GeoServer WMTS",
                            "autoload": false
                        }
                    },
                    "staticServices": {
                        "default_map_backgrounds": {
                            "type": "backgrounds",
                            "title": "Default Backgrounds",
                            "titleMsgId": "defaultMapBackgroundsServiceTitle",
                            "autoload": true,
                            "backgrounds": [
                                {
                                    "type": "osm",
                                    "title": "Open Street Map",
                                    "name": "mapnik",
                                    "source": "osm",
                                    "group": "background"
                                },
                                {
                                    "type": "tileprovider",
                                    "title": "NASAGIBS Night 2012",
                                    "provider": "NASAGIBS.ViirsEarthAtNight2012",
                                    "name": "Night2012",
                                    "source": "nasagibs",
                                    "group": "background"
                                },
                                {
                                    "type": "tileprovider",
                                    "title": "OpenTopoMap",
                                    "provider": "OpenTopoMap",
                                    "name": "OpenTopoMap",
                                    "source": "OpenTopoMap",
                                    "group": "background"
                                },
                                {
                                    "format": "image/jpeg",
                                    "group": "background",
                                    "name": "s2cloudless:s2cloudless",
                                    "opacity": 1,
                                    "title": "Sentinel 2 Cloudless",
                                    "type": "wms",
                                    "url": "https://1maps.geo-solutions.it/geoserver/wms",
                                    "source": "s2cloudless",
                                    "singleTile": false
                                },
                                {
                                    "source": "ol",
                                    "group": "background",
                                    "title": "Empty Background",
                                    "fixed": true,
                                    "type": "empty"
                                }
                            ]
                        }
                    }
                }
            },
            "locales": {
                "supportedLocales": {
                    "it": {
                        "code": "it-IT",
                        "description": "Italiano"
                    },
                    "en": {
                        "code": "en-US",
                        "description": "English"
                    },
                    "fr": {
                        "code": "fr-FR",
                        "description": "Français"
                    },
                    "de": {
                        "code": "de-DE",
                        "description": "Deutsch"
                    },
                    "es": {
                        "code": "es-ES",
                        "description": "Español"
                    }
                }
            }
        }
    }
}
```

File: web/client/components/plugins/enhancers/withModulePlugins.js
Comment (allyoucanmap) [inline web/client/components/plugins/enhancers/withModulePlugins.js:17]: please add the jsdoc for this file

File: web/client/containers/Page.jsx
Comment (allyoucanmap) [reply web/client/containers/Page.jsx:19]: please remove the compose function is not needed, connect is enough

File: web/client/hooks/useModulePlugins.js
Comment (allyoucanmap) [inline web/client/hooks/useModulePlugins.js:30]: please add jsdoc for this file

File: web/client/utils/ModulePluginsUtils.js
Comment (allyoucanmap) [inline web/client/utils/ModulePluginsUtils.js:13]: please add jsdoc for all the functions in this file

File: web/client/utils/StateUtils.js
Comment (allyoucanmap) [inline web/client/utils/StateUtils.js:228]: If we are removing the augmentStore we need to add a note in the migration guideline doc explaining the alternative with the new storeManager because this is a breaking change

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

**Comment by tdipisa** [comment]
@alexander-fedorenko with the exception of changes @allyoucanmap requested above that should be provided asap, below is the order of priority of thing that would like to do before merging this PR:

1. Setup the lazy loading also for ThematicLayerPlugin and StyleEditorPlugin
2. Verify if the configuration for Catalog, Measure and Annotations is still valid if the plugins are loaded after the initialization of standard app
3. Verify if we can provide a lazy loading of _css-tree_ 

With the above three point we should be in a really good state in term of performances. Early after this we will check how to proceed to schedule the others:

> -  ContextPlugin - no data is visible on context page. This is a flow issue, required action or actions are dispatched before plugin epics are registered and ready to listen for actions.
> -  Dashboard - same as above.
> -  DashboardsPlugin - same as above.
> -  GeoStoryPlugin - same as above.
> -  GeoStoriesPlugin - same as above.
> -  MapsPlugin - same as above.
> -  FeedbacMaskPlugin - rely on INIT_MAP action that is dispatched before plugin epics are registered.
> -  PrintPlugin - probably a conflict with how plugin is defined. Need to be investigated.
> -  UserSession - need investigation.

File: web/client/utils/StateUtils.js
Comment (allyoucanmap) [inline web/client/utils/StateUtils.js:209]: Just tested the mute and unmute functionality for epics and it is not working and the reason seems related to the key used in the muteState global object cache because the key sometimes contain the `Plugin` suffix and sometimes not. The only solution I came up is to trim the end always but maybe there is a better one:

```js
const pluginName = trimEnd(key, 'Plugin');
muteState[pluginName]
```

With the correct key it seems to work for common epic such as

```js
const msTestEpic = (action$) => {
    return action$.ofType('@@router/LOCATION_CHANGE')
        .switchMap(() => {
            console.log('TEST');
            return Observable.empty();
        });
};
```

while epic that has a subprocess that does not depend on the action (eg timer) are not "muted" (probably they are muted but the subprocess does not know if it should stop or not). I don't think these is a common use case but it could happen that you need a polling stream in a extension/plugin

```js
const msTestEpicTimer = (action$) => {
    return action$.ofType('@@router/LOCATION_CHANGE')
        .switchMap(() => {
            return Observable.interval(100)
                .switchMap(() => {
                    console.log('TEST');
                    return Observable.empty();
                });
        });
};
```

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

**Comment by tdipisa** [comment]
Thank you so much @allyoucanmap.
@alexander-fedorenko please let us know for the above that is also a quite important thing. So, in order of priority I would say:

1. Setup the lazy loading also for ThematicLayerPlugin and StyleEditorPlugin
2. Verify the mute and unmute functionality for epics
3. Verify if the configuration for Catalog, Measure and Annotations is still valid if the plugins are loaded after the initialization of standard app
4. Verify if we can provide a lazy loading of css-tree

File: web/client/utils/StateUtils.js
Comment (alexander-fedorenko) [reply web/client/utils/StateUtils.js]: Here is the working snippet that actually turns off substream whenever isActive$ become false. Basically, code is the same as for the outer wrapper of each isolatedEpic. With the recent commit both modulePlugins epics and appEpics are wrapped wtih isolateEpics, so all of them will have third argument available, even though it will have no effect on appEpics since there is no code to mute them.
```js
export const dummyEpic = (action$, store, isActive$) => action$.ofType(ACTION)
    .switchMap(() => {
        return Rx.Observable.interval(1000)
            .let(semaphore(isActive$.startWith(true)))
            .switchMap(() => {
                console.log('TEST');
                return Rx.Observable.empty();
            });
    });
```

**Comment by alexander-fedorenko** [comment]
1. Added dynamic import for ThematicLayerPlugin and StyleEditorPlugin. I can't test them locally, but they have no epics relying on actions triggered before PluginsContainer get rendered, so both of them should be fine.
2. Applied corrections to the code that mute or unmute epics. Now they will normalize plugins names just like it happens in plugins cache.
3. Configuration for Catalog, Measurement, Annotations is not loaded properly unless reducers are added into list of appReducers. In case of dynamic import default reducer state overwrite configuration passed upon store creation. So either they have to be in appReducers, or default state in reducer should be taken somewhere outside using utility function.
4. We can use dynamic import, checking with @allyoucanmap if suggested approach is sufficient.

Test coverage is extended, `withModulePlugins` & `useModulePlugins` are covered.

Two updates are on the way:
- Documentation about transforming extension into module extension.
- Making sub-stream mutable when outer epic is muted.

**Comment by alexander-fedorenko** [comment]
## Bundle analyzer after recent updates


### Static plugins included
### Parsed
image

### Gzipped
image

### Static plugins excluded

### Parsed
image


### Gzipped
image

**Comment by allyoucanmap** [comment]
@tdipisa @alexander-fedorenko 

> 1. Added dynamic import for ThematicLayerPlugin and StyleEditorPlugin. I can't test them locally, but they have no epics relying on actions triggered before PluginsContainer get rendered, so both of them should be fine.

We will test this once merged

> 2. Applied corrections to the code that mute or unmute epics. Now they will normalize plugins names just like it happens in plugins cache.

ok, good

> 3. Configuration for Catalog, Measurement, Annotations is not loaded properly unless reducers are added into list of appReducers. In case of dynamic import default reducer state overwrite configuration passed upon store creation. So either they have to be in appReducers, or default state in reducer should be taken somewhere outside using utility function.

I think we can keep them as appReducers, they haven't effect on app size,
checking the Analyzer results latest branch update is 6.5 MB vs 6.75 MB previous analysis (not Gzipped)

We will open a separate issue to solve the configuration for these plugins

> 4. We can use dynamic import, checking with @allyoucanmap if suggested approach is sufficient.

We will open a separate issue to investigate properly on the css-tree import


> Test coverage is extended, `withModulePlugins` & `useModulePlugins` are covered.
> 
> Three updates are on the way:
> - Documentation about transforming extension into module extension.
> - Guideline to update from augmentStore to storeManager use.
> - Making sub-stream mutable when outer epic is muted.

good, waiting for this latest update for the final review, thanks

**Comment by tdipisa** [comment]
Wonderful @allyoucanmap and @alexander-fedorenko. @alexander-fedorenko I'm looking forward to receiving the last update soon

**Comment by alexander-fedorenko** [comment]
@tdipisa @allyoucanmap Documentation update is committed. There are two new documentation chapters.
I also recovered augmentStore in StateUtils but marked it as deprecated in favor of storeManager usage.

**Review by allyoucanmap — APPROVED** [review]

**Comment by tdipisa** [comment]
@ElenaGallo let's schedule a call to discuss about this and schedule the testing phase.

**Comment by tdipisa** [comment]
@ElenaGallo after the test phase the update provided through this PR will need to be backported to 2022.02.xx to be included in the release 2022.02.00

**Comment by offtherailz** [comment]
> ## Bundle analyzer after recent updates
> ### Static plugins included
> ### Parsed
> image
> 
> ### Gzipped
> image
> 
> ### Static plugins excluded
> ### Parsed
> image
> 
> ### Gzipped
> image

Just a note about this:
- Most of the libs in the main package can be optimized more in the future : 
    - QuillJS should be replaced with  React-draft-wysiwyg (custom)
    - React-draft-wysiwyg (and so draft.js) should be loaded with React.lazy in every instance. 
    - `pdf.js` should be already loaded in a lazy way. For some reason is included in the main package again. Same for shp.js or zip.js. Anyway they look a lot small, so we may need to investigate more about these libs parts. 
- From lighthouse analysis, to increase the score we may need to imrpove caching of JS (expecially for the main JS) introducing the chunk hash in the name (e.g. `mapstore2.js` --> `mapstore2-aaasddgfdgt43.js`)  Anyway the performance looks to be  2.5 - 3 times better. 

Summarizing, `pdf.js` is the only thing that looks strange to me to see in the main package. The rest of the things can be improved one by one later.

**Pull request merged**
